db/src/lookup_list.rs

```diff
@@ -0,0 +1,145 @@
+use crate::table::Table;
+use crate::{DbResult, Logger, TableId};
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::hash::Hash;
+
+pub struct LookupList<K, V>
+where
+    K: Hash + Eq + Serialize,
+    V: Serialize + Eq,
+{
+    table_id: TableId,
+    inner: HashMap<K, Vec<V>>,
+    pub logger: Logger,
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum LogEntry<K, V> {
+    Push(K, V),
+    Remove(K, usize),
+    CreateKey(K),
+    ClearKey(K),
+    Clear,
+}
+
+impl<K, V> Table for LookupList<K, V>
+where
+    K: Hash + Eq + Serialize + DeserializeOwned,
+    V: Serialize + DeserializeOwned + Eq + Hash,
+{
+    fn init(table_id: TableId, logger: Logger) -> Self {
+        Self { table_id, inner: HashMap::default(), logger }
+    }
+
+    fn handle_event(&mut self, bytes: &[u8]) -> DbResult<()> {
+        match bincode::deserialize::<LogEntry<K, V>>(bytes)? {
+            LogEntry::Push(k, v) => {
+                self.push_inner(k, v);
+            }
+            LogEntry::Remove(k, idx) => {
+                if let Some(vec) = self.inner.get_mut(&k) {
+                    vec.remove(idx);
+                }
+            }
+            LogEntry::CreateKey(k) => {
+                self.inner.insert(k, Vec::new());
+            }
+            LogEntry::ClearKey(k) => {
+                self.inner.remove(&k);
+            }
+            LogEntry::Clear => {
+                self.inner.clear();
+            }
+        };
+
+        Ok(())
+    }
+
+    fn compact_repr(&self) -> DbResult<Vec<u8>> {
+        let mut repr = vec![];
+        for (k, values) in &self.inner {
+            if values.is_empty() {
+                let data = bincode::serialize(&LogEntry::<&K, &V>::CreateKey(k))?;
+                let mut data = Logger::log_entry(self.table_id, data);
+                repr.append(&mut data);
+                continue;
+            }
+            for v in values {
+                let data = bincode::serialize(&LogEntry::Push(k, v))?;
+                let mut data = Logger::log_entry(self.table_id, data);
+                repr.append(&mut data);
+            }
+        }
+
+        Ok(repr)
+    }
+}
+
+impl<K, V> LookupList<K, V>
+where
+    K: Hash + Eq + Serialize + DeserializeOwned,
+    V: Serialize + DeserializeOwned + Eq + Hash,
+{
+    pub(crate) fn push_inner(&mut self, k: K, v: V) -> bool {
```

Comment:
This fn should probably not return anything. In the other `insert_inner` I was propagating info up from `HashSet`'s `insert`, but here `push` doesn't return anything.